### PR TITLE
[REA-980] Json-schemas: add metachannel client events

### PR DESCRIPTION
**Pull request by Jmgr**
This adds metachannel client events to the Json schemas.

### CODE REVIEW

**Comment by Jmgr** [comment]
See comment https://ably.atlassian.net/browse/REA-980?focusedCommentId=24385

File: json-schemas/src/client-events-api-requests.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "Ably API Requests Client Event",
+  "type": "object",
```
Comment (lmars) [inline json-schemas/src/client-events-api-requests.json:4]: Let's also add a description, for example:

```
"description": "A message sent on the clientEvents:apiRequests metachannel describing a sampled client api request."
```

File: json-schemas/src/client-events-api-requests.json [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  "properties": {
+    "host": {
+      "type": "string",
+      "description": "the host the request was made to"
```
Comment (lmars) [inline json-schemas/src/client-events-api-requests.json:8]: Let's format these descriptions as proper sentences (i.e. start with a capital letter, and end with a full stop).

Also, I assume this is the HTTP Host header from the request, in which case I'd alter the description to:

```
The HTTP Host header of the API request.
```

File: json-schemas/src/client-events-api-requests.json [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    },
+    "requestId": {
+      "type": "string",
+      "description": "the unique ID of the request that Ably can use to correlate a request event with internal logs, if necessary"
+    },
+    "region": {
+      "type": "string",
+      "description": "the region the request was made in"
```
Comment (lmars) [inline json-schemas/src/client-events-api-requests.json:16]:
```
The region the request was processed in.
```

File: json-schemas/src/client-events-api-requests.json [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    },
+    "headers": {
+      "type": "object",
+      "description": "the headers sent with the request"
```
Comment (lmars) [inline json-schemas/src/client-events-api-requests.json:20]:
```
The HTTP headers of the API request.
```

File: json-schemas/src/client-events-api-requests.json (hunk 3 above)
Comment (lmars) [reply json-schemas/src/client-events-api-requests.json:16]: And I just looked at a sample message, and it seems this is in fact a site, not a region (i.e. it has a `-A` suffix):

```
"region": "eu-central-1-A"
```

Should we change it to `site` now before we publish this schema?

File: json-schemas/src/client-events-api-requests.json (hunk 4 above)
Comment (lmars) [reply json-schemas/src/client-events-api-requests.json:20]: Also, it's worth noting that the HTTP header keys are normalised to all lower case (i.e. `user-agent` rather than `User-Agent`)

File: json-schemas/src/client-events-api-requests.json (hunk 2 above)
Comment (Jmgr) [reply json-schemas/src/client-events-api-requests.json:8]: > Let's format these descriptions as proper sentences (i.e. start with a capital letter, and end with a full stop).

For some reason I was convinced that the convention for JSON Schema was to drop both, but I cannot find any evidence of that now. 

> Also, I assume this is the HTTP Host header from the request, in which case I'd alter the description to:

Could be, I'll check.

File: json-schemas/src/client-events-api-requests.json (hunk 3 above)
Comment (Jmgr) [reply json-schemas/src/client-events-api-requests.json:16]: That would make sense, I'll also change the code in realtime to do that.

File: json-schemas/src/client-events-api-requests.json (hunk 4 above)
Comment (Jmgr) [reply json-schemas/src/client-events-api-requests.json:20]: Yes. Do you mean that should be included in the description? Or that the code should enforce that by making all headers lower case?

File: json-schemas/src/client-events-api-requests.json (hunk 2 above)
Comment (lmars) [reply json-schemas/src/client-events-api-requests.json:8]: My reasoning is that I expect these descriptions to appear in both documentation (e.g. [here](https://ably.com/documentation/realtime/metachannels#connections) and in UIs when users interact with these messages, where I would expect them to be proper sentences.

File: json-schemas/src/client-events-api-requests.json (hunk 4 above)
Comment (lmars) [reply json-schemas/src/client-events-api-requests.json:20]: I think it should be in the description to inform the user, but no enforcement necessary since HTTP header names are case insensitive.

File: json-schemas/src/client-events-api-requests.json [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+      "type": "string",
+      "description": "The unique ID of the request that Ably can use to correlate a request event with internal logs, if necessary."
+    },
+    "site": {
+      "type": "string",
+      "description": "The site the request was processed in."
+    },
+    "headers": {
+      "type": "object",
+      "description": "The HTTP headers sent with the request. Keys should be lower case."
```
Comment (lmars) [inline json-schemas/src/client-events-api-requests.json:21]: "should" is too strong, it's not a requirement, it's just the convention we use:

```
Header names are lower case (e.g. 'accept', 'content-type', 'user-agent', etc.)
```

File: json-schemas/src/client-events-api-requests.json [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    },
+    "query": {
+      "type": "object",
+      "description": "The details of the parsed request query, including information such as the request format."
+    },
+    "path": {
+      "type": "string",
+      "description": "The path of the endpoint called in the request, for example /channels/{channel-name}/messages."
```
Comment (lmars) [inline json-schemas/src/client-events-api-requests.json:29]: "query" and "path" should probably have similar descriptions, they currently feel a bit wordy and inconsistent. How about:

- query: "The parsed query component of the API request URL (e.g. {\"format\":\"msgpack\"})."
- path: "The path component of the API request URL (e.g. /channels/{channel-name}/messages)."

File: json-schemas/src/client-events-api-requests.json [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    },
+    "channels": {
+      "type": "array",
+      "description": "A list of channels the request was made against."
```
Comment (lmars) [inline json-schemas/src/client-events-api-requests.json:33]: I don't like the term "made against", how about "references"?

"The list of channels the API request references."

File: json-schemas/src/client-events-api-requests.json [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    },
+    "error": {
+      "type": "object",
+      "description": "The details of any error encountered whilst making the request. It includes an error message, error code and HTTP status code."
```
Comment (lmars) [inline json-schemas/src/client-events-api-requests.json:37]: Given this error object has a well defined structure (as indicated here in the description), should it have a schema? Maybe it should be a separate schema that's referenced here?

File: json-schemas/src/client-events-connections.json [hunk 9]
```diff
@@ -0,0 +1,50 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "Ably Connections Client Event",
+  "type": "object",
+  "description": "A message sent on the clientEvents:connections metachannel describing a sampled client connection.",
+  "properties": {
+    "host": {
+      "type": "string",
+      "description": "The host the connection was made to."
```
Comment (lmars) [inline json-schemas/src/client-events-connections.json:9]: This should be consistent with the descriptions in the api requests schema.

File: json-schemas/src/client-events-api-requests.json (hunk 8 above)
Comment (Jmgr) [reply json-schemas/src/client-events-api-requests.json:37]: > Given this error object has a well defined structure (as indicated here in the description), should it have a schema?

Good point.

> Maybe it should be a separate schema that's referenced here?

Using an external schema will be painful to implement in unit tests, so I'd rather be for adding the three missing properties there.

File: json-schemas/src/client-events-connections.json [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+      "type": "string",
+      "description": "The HTTP Host header sent with the connection."
+    },
+    "requestId": {
+      "type": "string",
+      "description": "The unique ID of the request that Ably can use to correlate a connection event with internal logs, if necessary."
+    },
+    "site": {
+      "type": "string",
+      "description": "The site the connection was made to."
```
Comment (lmars) [inline json-schemas/src/client-events-connections.json:17]: This should be the same as the description in the api requests schema.

File: json-schemas/src/client-events-connections.json [hunk 11]
```diff
@@ -0,0 +1,65 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "Ably Connections Client Event",
+  "type": "object",
+  "description": "A message sent on the clientEvents:connections metachannel describing a sampled client connection.",
+  "properties": {
+    "host": {
+      "type": "string",
+      "description": "The HTTP Host header sent with the connection."
```
Comment (lmars) [inline json-schemas/src/client-events-connections.json:9]: You don't really _send_ a connection. How about:

"The HTTP Host header used when establishing the connection."

(and similarly in other descriptions, replace "making" or "sending" with "establishing" a connection)

Comment (Jmgr) [reply json-schemas/src/client-events-connections.json:9]: I thought it was implied that this meant "sent with the connection request", but I'm ok with your suggestion.

**Review by lmars — APPROVED** [review]
